File: android/src/main/kotlin/lazebny/io/jc/logic/JcConferenceStateApiImpl.kt
package lazebny.io.jc.logic

import lazebny.io.jc.common.JcConferenceStateApi
import lazebny.io.jc.logic.JcWrapper.JCManager

class JcConferenceStateApiImpl : JcConferenceStateApi {
    override fun microphone(): Boolean {
        val isAudio = JCManager.getInstance().mediaChannel.selfParticipant?.isAudio ?: false
        val isAudioStart = JCManager.getInstance().mediaDevice.isAudioStart
        return isAudio && isAudioStart
    }

    override fun video(): Boolean {
        val isVideo = JCManager.getInstance().mediaChannel.selfParticipant?.isVideo ?: false
        val isCameraOpen = JCManager.getInstance().mediaDevice.isCameraOpen
        return isVideo && isCameraOpen
    }

    override fun speaker(): Boolean {
        return JCManager.getInstance().mediaDevice.isSpeakerOn
    }

    override fun uid(): String {
        return JCManager.getInstance().mediaChannel?.selfParticipant?.userId ?: ""
    }

    override fun conferenceStatus(): String {
        val participants = JCManager.getInstance()?.mediaChannel?.participants
        if (participants.isNullOrEmpty()) {
            return "off"
        }

        if (participants.size == 1) {
            return "waiting"
        }

        return "on"
    }

    override fun members(): List<Map<String, *>> {
        val participants = JCManager.getInstance()?.mediaChannel?.participants
        if (participants.isNullOrEmpty()) {
            return emptyList()
        }

        val participantsWithoutSelf = participants.toMutableList()
        participantsWithoutSelf.remove(JCManager.getInstance().mediaChannel.selfParticipant)

        return participantsWithoutSelf.map {
            mapOf(
                "uid" to it.userId,
                "name" to it.displayName,
                "microphone" to it.isAudio,
                "video" to it.isVideo,
            )
        }
    }
}

File: ios/Classes/logic/JcConferenceStateApiImpl.swift
//
//  JcConferenceStateApi.swift
//  jc
//
//  Created by Михаил on 31.05.2023.
//

import Foundation
import JCSDKOC

class JcConferenceStateApiImpl : JcConferenceStateApi {
    func microphone() -> Bool {
        let audio = JCRoom.shared.mediaChannel.selfParticipant?.audio ?? false
        let audioStart = JCRoom.shared.mediaDevice.audioStart
        return audio && audioStart
    }
    
    func video() -> Bool {
        let video = JCRoom.shared.mediaChannel.selfParticipant?.video ?? false
        let cameraOpen = JCRoom.shared.mediaDevice.cameraOpen
        return video && cameraOpen
    }
    
    func uid() -> String {
        return JCRoom.shared.mediaChannel.selfParticipant?.userId ?? ""
    }
    
    func speaker() -> Bool {
        return JCRoom.shared.mediaDevice.isSpeakerOn()
    }
    
    func members() -> [[String : Any]] {
        let participants = JCRoom.shared.mediaChannel.participants
        if (participants.isEmpty) {
            return []
        }
        
        var participantsWithoutSelf = participants.compactMap({ ($0 as! JCMediaChannelParticipant) })
        if let selfParticipant = JCRoom.shared.mediaChannel.selfParticipant,
           let index = participantsWithoutSelf.firstIndex(where: { $0 == selfParticipant }) {
            participantsWithoutSelf.remove(at: index)
        }

        return participantsWithoutSelf.map { participant in
            [
                "uid": participant.userId!,
                "name": participant.displayName!,
                "microphone": participant.audio,
                "video": participant.video,
            ]
        }
    }
    
    func conferenceStatus() -> String {
        let participants = JCRoom.shared.mediaChannel.participants
        if (participants.isEmpty) {
            return "off"
        }
        
        if (participants.count == 1) {
            return "waiting"
        }
        
        return "on"
    }
    
    
}
